primitives/messages/src/target_chain.rs

```diff
@@ -100,7 +100,17 @@ pub trait MessageDispatch<Fee> {
 	///
 	/// It is up to the implementers of this trait to determine whether the message
 	/// is invalid (i.e. improperly encoded, has too large weight, ...) or not.
-	fn dispatch(message: DispatchMessage<Self::DispatchPayload, Fee>);
+	///
+	/// If your configuration allows paying dispatch fee at the target chain, then
+	/// it must be paid inside this method to the `relayer_account`.
+	///
+	/// Returns 'unused' dispatch weight that will be deducted from total delivery transaction
+	/// weight, thus reducing the transaction cost. This shall not be zero in two cases:
+	///
+	/// 1) if message has been dispatched successfully, but post-dispatch weight is less than
+	///    the weight, declared by the message sender;
+	/// 2) if message has not been dispatched at all.
```

Comment:
I was thinking a bit about burning part of that weight, or returning part of the fee to the message sender, but I concluded that it doesn't make much sense for now. Leaving the comment here for posterity though.

1. In case of target-chain fee payment, messages senders are heavily incentivised to specify the lowest possible weight, cause they pay up-front. Any unused weight is benefiting the relayers now by increasing their reward for accepting target-chain fee payment, which inherently increases the risk for them. So that seems to be quite aligned.
2. In case of source-chain fee payment, the situation is quite similar - in such case relayers risk is increased by price fluctuation though.
3. Hence message senders does not really benefit directly from having the dispatch consume less weight than declared, cause they pay for dispatch upfront. Burning part of the unused weight would only reduce reward for the relayers (which are exposed to a bunch of risks already) and would unnecessarily inflate block consumption.
4. Returning part of unused fee to senders reduces their incentive to calculate the fee correctly and increases risk for the relayers.

Overall we might consider doing a more detailed analysis, perhaps there is some fair split between returning/burning/leaving to the relayer, but taking into account real-world data will be a great help for this, so I'd rather collect it first :)
My gut feeling is that relayers are more on the upside here, but I think it's acceptable for the first deployment.